examples/research_projects/vqgan-clip/VQGAN_CLIP.py

```diff
@@ -0,0 +1,266 @@
+
+        self.iterations = iterations
+        self.lr = lr
+        self.log = log
+        self.make_grid = make_grid
+        self.return_val = return_val
+        self.quantize = quantize
+        self.latent_dim = self.vqgan.decoder.z_shape
+
+    def make_animation(self, input_path=None, output_path=None, total_duration=5, extend_frames=True):
+        """
+        Make an animation from the intermediate images saved during generation.
+        By default, uses the images from the most recent generation created by the generate function.
+        If you want to use images from a different generation, pass the path to the folder containing the images as input_path.
+        """
+        images = []
+        if output_path is None:
+            output_path = "./animation.gif"
+        if input_path is None:
+            input_path = self.save_path
+        paths = list(sorted(glob(input_path + "/*")))
+        assert len(paths), print(
+            "No images found in save path, aborting (did you pass save_intermediate=True to the generate function?)"
+        )
+        if len(paths) == 1:
+            print("Only one image found in save path, (did you pass save_intermediate=True to the generate function?)")
+        frame_duration = total_duration / len(paths)
+        durations = [frame_duration] * len(paths)
+        if extend_frames:
+            durations[0] = 1.5
+            durations[-1] = 3
+        for file_name in paths:
+            if file_name.endswith(".png"):
+                images.append(imageio.imread(file_name))
+        imageio.mimsave(output_path, images, duration=durations)
+        print(f"gif saved to {output_path}")
+
+    def _get_latent(self, path=None, img=None):
+        assert path or img, "Input either path or tensor"
+        if img is not None:
+            raise NotImplementedError
+        x = preprocess(Image.open(path), target_image_size=256).to(self.device)
+        x_processed = preprocess_vqgan(x)
+        z, *_ = self.vqgan.encode(x_processed)
+        return z
+
+    def _add_vector(self, transform_vector):
+        """Add a vector transform to the base latent and returns the resulting image."""
+        base_latent = self.latent.detach().requires_grad_()
+        trans_latent = base_latent + transform_vector
+        if self.quantize:
+            z_q, *_ = self.vqgan.quantize(trans_latent)
+        else:
+            z_q = trans_latent
+        return self.vqgan.decode(z_q)
+
+    def _get_clip_similarity(self, prompts, image, weights=None):
+        clip_inputs = self.clip_preprocessor(text=prompts, images=image, return_tensors="pt", padding=True)
+        clip_outputs = self.clip(**clip_inputs)
+        similarity_logits = clip_outputs.logits_per_image
+        if weights is not None:
+            similarity_logits = similarity_logits * weights
+        return similarity_logits.sum()
+
+    def _get_CLIP_loss(self, pos_prompts, neg_prompts, image):
```

Comment:
To be consistent with naming above

```suggestion
    def _get_clip_loss(self, pos_prompts, neg_prompts, image):
```